src/main.rs

```diff
@@ -1,20 +1,151 @@
-use std::net::TcpListener;
-use std::thread::spawn;
-
-use tungstenite::accept;
-
-fn main() {
-    let server = TcpListener::bind("127.0.0.1:3012").unwrap();
-    for stream in server.incoming() {
-        spawn(move || {
-            let mut websocket = accept(stream.unwrap()).unwrap();
-
-            loop {
-                let msg = websocket.read_message().unwrap();
-                if msg.is_binary() || msg.is_text() {
-                    websocket.write_message(msg).unwrap();
-                }
-            }
-        });
+use std::{
+    collections::HashMap,
+    fs,
+    io::Error as IoError,
+    net::SocketAddr,
+    str,
+    sync::{Arc, Mutex},
+};
+
+use futures::prelude::*;
+use futures::{
+    channel::mpsc::{unbounded, UnboundedSender},
+    future, join, pin_mut,
+};
+
+use async_std::net::{TcpListener, TcpStream};
+use async_std::task;
+use tungstenite::protocol::Message;
+
+type Tx = UnboundedSender<Message>;
+type PeerMap = Arc<Mutex<HashMap<SocketAddr, Tx>>>;
+
+async fn handle_websocket_connection(
+    peer_map: PeerMap,
+    raw_stream: TcpStream,
+    addr: SocketAddr,
+    id: u8,
+) {
+    println!("Incoming TCP connection from: {}", addr);
+
+    let ws_stream = async_tungstenite::accept_async(raw_stream)
+        .await
+        .expect("Error during the websocket handshake occurred");
+    println!("WebSocket connection established: {}", addr);
+
+    let (tx, rx) = unbounded();
+
+    // Alert client of their ID.
+    tx.unbounded_send(Message::Text(format!("id:{}", id)))
+        .unwrap();
+
+    // Insert the write part of this peer to the peer map.
+    peer_map.lock().unwrap().insert(addr, tx);
+
+    let (outgoing, incoming) = ws_stream.split();
+
+    let broadcast_incoming = incoming.try_for_each(|msg| {
+        // println!("{} | {:?}", addr, msg);
+        if msg.is_close() { return future::ok(()) }
```

Comment:
I added this so that when one client closed their socket, it wasn't broadcast to the other clients closing all their sockets. Not sure if this is a good idea. Going to ask on the `async-tungstenite` GH